src/gui/scheduling/syncscheduler.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) by Hannah von Reth <hannah.vonreth@owncloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "gui/scheduling/syncscheduler.h"
+
+#include "gui/folderman.h"
+#include "gui/scheduling/etagwatcher.h"
+#include "libsync/configfile.h"
+
+using namespace std::chrono_literals;
+
+using namespace OCC;
+
+SyncScheduler::SyncScheduler(FolderMan *parent)
+    : QObject(parent)
+    , _etagManager(new EtagWatcher(parent, this))
+{
+    // Normal syncs are performed incremental but when fullLocalDiscoveryInterval times out
+    // a complete local discovery is performed.
+    // This timer here triggers a sync independent of etag changes on the server.
+    auto *fullLocalDiscoveryTimer = new QTimer(this);
+    fullLocalDiscoveryTimer->setInterval(ConfigFile().fullLocalDiscoveryInterval() + 2min);
+    connect(fullLocalDiscoveryTimer, &QTimer::timeout, this, [parent, this] {
+        for (auto *f : parent->folders()) {
+            if (f->isReady() && f->accountState()->state() == AccountState::State::Connected) {
+                enqueueFolder(f);
+            }
+        }
+    });
+}
+
+void SyncScheduler::enqueueFolder(Folder *folder, Priority priority)
+{
+    Q_ASSERT(folder->isReady());
+    Q_ASSERT(folder->canSync());
+    auto old = _scheduledFolders.find(folder);
+    if (old == _scheduledFolders.cend()) {
+        // the folder is not yet scheduled
+        _queue.emplace(folder, priority);
+    } else {
+        _scheduledFolders.insert({folder, priority});
+        // if the new priority is higher we need to rebuild the queue
+        if (priority > old->second) {
```

Comment:
This code should never have landed here. We need a standalone class for the raw priority queue.

Please remind me why we couldn't use `std::priority_queue`. (This could go into a comment by the way.)

Reply:
We do use std::priority_queue but a priority queue but we use a pair and could queue the same folder multiple times.